### PR TITLE
Add cpuset statistics

### PR DESCRIPTION
**Pull request by piowag**
Adds statistics for cpuset cgroup

Signed-off-by: Piotr Wagner <piotr.wagner@intel.com>

### CODE REVIEW

**Comment by kolyshkin** [comment]
@piowag you need to squash your commits

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:63]: you can avoid this check and go right to splitting. If `len(splitted) == 1` it means there is no dash.

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:64]: nit: `strings.Split(s, "-", 3)` so you won't split more than necessary. We use `3` to detect extra `-`.

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:77]: Perpahs `invalid values in %s`?

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:83]: Maybe move the single value case up there and return early for better readability?

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:53]: The name is way too generic. Perhaps `getCpusetStat` or something like that?

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:98]: You can avoid declaring it here if you modify `getNumericStat()` to return in. IOW
```go
cpus, err := getNumericStat(path, "cpuset.cpus")
```

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset_test.go:440]: Perhaps you can combine all these tests into one, by having an initial set of files written, and checking if the values read are as expected, and then checking all the error cases like this
```go
        testCases := []struct {
                desc string
                filename, contents string
                removeFilename string
        }{
                "missing SchedRelaxDomainLevelFile",
                removeFilename = "cpuset.sched_relax_domain_level"
        }

for _, tc := range errTestCases {
￼	cpuset := &CpusetGroup{}
￼	actualStats := *cgroups.NewStats()
￼	err := cpuset.GetStats(helper.CgroupPath, &actualStats)
￼	if err == nil {
￼		t.Errorf("case %s: expected failure, got nil", tc.desc)
￼	}
}
```

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:141]: Are you sure you want to return an error if any of these files do not exist? What if some older kernel do not provide some?

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: 2 or 3 could also mean no dash

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go:72]: If it can perform more than one split in this place, it means that file is corrupted and should error out.

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Applied

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Applied

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Applied

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Applied

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset_test.go]: Condensed them into two tests: positive case and all negative cases

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Changed stat gathering method to try to gather as much as possible instead of quitting at first error

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpuset.go:74]: right, that is what I meant.
1. Use `strings.SplitN(s, "-", 3)` instead of `strings.Split(s, "-")` to avoid potential extra splitting that you don't need.
2. Since you use `3` you can still detect the corrupted input by checking for `len(splitted) != 2`

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:76]: error messages are often wrapped, so use lowercase (`s/Invalid/invalid/`)

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:87]: ditto

File: libcontainer/cgroups/fscommon/utils.go
Comment (kolyshkin) [inline libcontainer/cgroups/fscommon/utils.go:86]: maybe `s/string(contents)/trimmed/` since you're actually parsing `trimmed`.

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset_test.go:238]: you can avoid piling up errors if you use `t.Errorf()` instead of `t.Fatalf()`.

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:104]: I think it does not make sense to try all the files and then still return an error. I mean, you can just return the first error you have encountered (unless I am missing something here).

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpuset.go:74]: @piowag ^^^

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpuset.go:76]: @piowag ^^^

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpuset.go:87]: @piowag ^^^

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpuset.go:104]: @piowag please let me know what do you think about the above suggestion

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpuset_test.go:238]: @piowag ^^^

File: libcontainer/cgroups/fscommon/utils.go
Comment (kolyshkin) [reply libcontainer/cgroups/fscommon/utils.go:86]: @piowag ^^^

**Comment by kolyshkin** [comment]
@piowag please let us know if you still can / want to work on this one

File: libcontainer/cgroups/fscommon/utils.go
Comment (piowag) [reply libcontainer/cgroups/fscommon/utils.go:86]: Should this also be applied to `GetCgroupParamUint`

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go:104]: Returning all errors seems better for user trying to debug the issue.

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Applied

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Applied

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset_test.go]: Applied

**Comment by piowag** [comment]
Sorry, I was absent a little. Yes I still can and want to work on this one.

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset.go:66]: I think it would be more optimal to do something like this

```go
splitted := strings.SplitN(s, "-", 3)
switch len(splitted) {
  case 3:
    return extracted, fmt.Errorf("invalid values in %s", filepath.Join(path, filename))
  case 2:
    // treat as min, max
  case 1:
    value, err := strconv.ParseUint(s, 10, 16)
}
```

This way you prevent the double-scanning of a string for `-`

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset_test.go:48]: You are introducing a global variable named `testCases` here and you only use it in a single function. I think this should either be renamed or moved to that function.

Comment (piowag) [reply libcontainer/cgroups/fs/cpuset_test.go]: renamed

File: libcontainer/cgroups/fs/cpuset.go
Comment (piowag) [reply libcontainer/cgroups/fs/cpuset.go]: Applied

**Comment by kolyshkin** [comment]
One thing I don't like about this patch is it requires all the cpuset files being present. On my system (kernel 5.6) I do not have, for example, `cpuset.mem_hardwall`. Does that mean that `GetStats()` will now return an error?

For example, `(* BlkioGroup).GetStat` ignores the ENOENT, returning no errors if some files are not present.

**Comment by piowag** [comment]
Ignoring ENOENT for all files, but cpuset.cpus and cpuset.mems which are mandatory.

**Comment by piowag** [comment]
@AkihiroSuda Could you take a look at this PR?

File: libcontainer/cgroups/fs/cpuset.go
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs/cpuset.go:167]: Can we use https://github.com/hashicorp/go-multierror ?

@kolyshkin @cyphar WDYT?

Comment (AkihiroSuda) [reply libcontainer/cgroups/fs/cpuset.go:167]: or https://github.com/uber-go/multierr

**Comment by AkihiroSuda** [comment]
@piowag WDYT about multierr?

**Comment by piowag** [comment]
Not sure if adding new 3-rd party library just to get rid of few lines is required. As for retaining the error type, it shouldn't be used anymore outside the logic of the `GetStats` function

**Review by AkihiroSuda — APPROVED** [review]

**Comment by piowag** [comment]
What do you think about changes introduced since your review @kolyshkin ?

**Comment by AkihiroSuda** [comment]
@kolyshkin PTAL?

File: libcontainer/cgroups/fs/cpuset.go
Comment (kwisniewski98) [inline libcontainer/cgroups/fs/cpuset.go:106]: In some cases `cpuset.cpus` and `cpuset.mems` might not exists, which (in my opinion) should not return an error

File: libcontainer/cgroups/fs/cpuset.go
Comment (thaJeztah) [inline libcontainer/cgroups/fs/cpuset.go:99]: Should these now be using `errors.Is()` ?

```suggestion
		if errors.Is(err, os.ErrNotExist) {
```

File: libcontainer/cgroups/fs/cpuset.go
Comment (thaJeztah) [inline libcontainer/cgroups/fs/cpuset.go:165]: Wondering if something like this would be somewhat cleaner (removing the intermediate variable, and nested `if`;

```suggestion
	stats.CpusetStats.SchedRelaxDomainLevel, err = fscommon.GetCgroupParamInt(path, "cpuset.sched_relax_domain_level")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
```

The whole function would then look something like;

```go
func (s *CpusetGroup) GetStats(path string, stats *cgroups.Stats) error {
	var (
		err        error
		errStrings []string
	)

	stats.CpusetStats.Cpus, err = getCpusetStat(path, "cpuset.cpus")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.CpuExclusive, err = fscommon.GetCgroupParamUint(path, "cpuset.cpu_exclusive")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.Mems, err = getCpusetStat(path, "cpuset.mems")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.MemHardwall, err = fscommon.GetCgroupParamUint(path, "cpuset.mem_hardwall")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.MemExclusive, err = fscommon.GetCgroupParamUint(path, "cpuset.mem_exclusive")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.MemoryMigrate, err = fscommon.GetCgroupParamUint(path, "cpuset.memory_migrate")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.MemorySpreadPage, err = fscommon.GetCgroupParamUint(path, "cpuset.memory_spread_page")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.MemorySpreadSlab, err = fscommon.GetCgroupParamUint(path, "cpuset.memory_spread_slab")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.MemoryPressure, err = fscommon.GetCgroupParamUint(path, "cpuset.memory_pressure")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.SchedLoadBalance, err = fscommon.GetCgroupParamUint(path, "cpuset.sched_load_balance")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}
	stats.CpusetStats.SchedRelaxDomainLevel, err = fscommon.GetCgroupParamInt(path, "cpuset.sched_relax_domain_level")
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		errStrings = append(errStrings, err.Error())
	}

	if len(errStrings) != 0 {
		return fmt.Errorf(strings.Join(errStrings, "\n"))
	}
	return nil
}
```

File: libcontainer/cgroups/fs/cpuset.go
Comment (thaJeztah) [reply libcontainer/cgroups/fs/cpuset.go:167]: Perhaps use `errors.New()` instead of `fmt.Errorf()` to preserve the stack

```suggestion
		return errors.New(strings.Join(errStrings, "\n"))
```

Comment (thaJeztah) [reply libcontainer/cgroups/fs/cpuset.go:167]: > Can we use https://github.com/hashicorp/go-multierror ?
> or https://github.com/uber-go/multierr

Looking at the code, nothing actually "handles" these errors, other than "something failed", so from that perspective I don't think a multi-error package would be needed (to preserve the actual errors?) https://github.com/opencontainers/runc/blob/9ada2e6d4f74989e965d37bf040a4130eba13439/libcontainer/container_linux.go#L195-L197

Wondering if this function should continue trying to collect all off these, and instead just return on the first error. Currently we're updating the `stats` struct that was passed, and try to fill in as many things as possible, but the results will be thrown away if any failure occurs anyway; see https://github.com/opencontainers/runc/blob/2cf8d240075dd322b9385100c9af4b149c973391/libcontainer/cgroups/systemd/v1.go#L331-L333

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM (not a maintainer)

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Ping @kolyshkin

**Comment by AkihiroSuda** [comment]
Sorry this needs rebase

**Comment by Creatone** [comment]
@kolyshkin @AkihiroSuda

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
@kolyshkin LGTY?

**Comment by AkihiroSuda** [comment]
@piowag Sorry, please rebase to retrigger CI

**Review by AkihiroSuda — DISMISSED** [review]

**Comment by AkihiroSuda** [comment]
@opencontainers/runc-maintainers PTAL. This has been open so long.

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset_test.go:212]: nit: no need for `\n` here.

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset_test.go:236]: nit: no need to use `fmt.Sprintf` here, as `t.Errorf` is perfectly capable of handling `%`-style arguments.

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs/cpuset_test.go:246]: ditto

File: libcontainer/cgroups/fscommon/utils.go
Comment (kolyshkin) [inline libcontainer/cgroups/fscommon/utils.go:75]: typo: `s/specidfied/specified/`

**Review by kolyshkin — CHANGES_REQUESTED** [review]
Left a few minor comments, overall LGTM

File: types/events.go
Comment (thaJeztah) [inline types/events.go:15]: should probably be "CPUSet" ?

```suggestion
	CPUSet            CPUSet              `json:"cpuset"`
```

File: types/events.go
Comment (thaJeztah) [inline types/events.go:74]:
```suggestion
type CPUSet struct {
```

File: types/events.go
Comment (thaJeztah) [inline types/events.go:76]:
```suggestion
	CPUExclusive          uint64   `json:"cpu_exclusive"`
```

File: types/events.go
Comment (thaJeztah) [inline types/events.go:75]:
```suggestion
	CPUs                  []uint16 `json:"cpus,omitempty"`
```

Not 100% sure if it should be lowercase or uppercase "s"

File: libcontainer/cgroups/stats.go
Comment (thaJeztah) [inline libcontainer/cgroups/stats.go:151]: Ideally

```suggestion
	CPUSetStats CPUSetStats `json:"cpuset_stats,omitempty"`
```

But I can see that being inconsistent with the other field; is that something that can still be changed, or would that break many things?

File: libcontainer/cgroups/stats.go
Comment (thaJeztah) [inline libcontainer/cgroups/stats.go:45]:
```suggestion
	CPUs []uint16 `json:"cpus,omitempty"`
```

File: libcontainer/cgroups/stats.go
Comment (thaJeztah) [inline libcontainer/cgroups/stats.go:47]:
```suggestion
	CPUExclusive uint64 `json:"cpu_exclusive"`
```

File: libcontainer/cgroups/fs/cpuset_test.go
Comment (thaJeztah) [reply libcontainer/cgroups/fs/cpuset_test.go:236]: Perhaps use sub-tests for these cases, then it would not be needed to include the name of the failed test in the error?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by kolyshkin** [comment]
@Creatone thanks for working on this! Can you please squash the commits?

**Review by kolyshkin — DISMISSED** [review]
LGTM

File: libcontainer/cgroups/fs/cpuset.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpuset.go]: This has been fixed

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
GitHub web ui says "Required statuses must pass before merging".

Seems to need to be rebased again to pick the recent CI changes, sorry

**Comment by AkihiroSuda** [comment]
@cyphar @kolyshkin Can we merge this to master directly without waiting for CI retrigger?

**Comment by Creatone** [comment]
@AkihiroSuda @kolyshkin

**Review by AkihiroSuda — APPROVED** [review]
Thanks!

File: events.go
Comment (thaJeztah) [inline events.go:145]: Looks like these types are compatible, and this could be simplified to;

```suggestion
	s.CPUSet = types.CPUSet(cg.CPUSetStats)
```

Comment (Creatone) [reply events.go]: Tricky :) Thanks!

**Review by AkihiroSuda — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM (not a maintainer)

thanks!

**Review by kolyshkin — APPROVED** [review]
LGTM

**Pull request merged**
